app/services/google_service/client_service.rb

```diff
@@ -0,0 +1,40 @@
+# frozen_string_literal: true
+
+module GoogleService
+  class ClientService
+    USER_AGENT = 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) '\
+                'AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.77 Safari/537.36'
+
+    BASE_SEARCH_URL2 = 'https://google.com/search'
+
+    def initialize(keyword, lang = 'en')
+      escaped_keyword = CGI.escape(keyword)
+      @uri = URI("#{BASE_SEARCH_URL2}?q=#{escaped_keyword}&hl=#{lang}&gl=#{lang}")
+    end
+
+    def call
```

Comment:
IMO with this implementation, what is becoming confusing is that value of instance variable result can be changed anywhere in any function, which can become difficult to track the value and reduces the readability of the code. Can you propose another implementation, maybe change the method name to `call!`, which can raise a error.

Reply:
@ankitkalia1195 
I've try to set `result` as scoped variable, thus preventing it to be access somewhere else. 

For the `call!`, I'm not sure: The ClientService is built so that errors are managed. In case an issue occurs, the method returns nil. Maybe I can add some comments to document the service? Or is the bang (!) still needed? :) 

Thanks!

Reply:
@malparty The reason exception can be raised here is, because you would be calling this function somewhere inside a background worker(sidekiq). Exception would automatically cause the worker to fail, and job could be automatically retried. Altough you are still logging the response, but the worker would not retry the job and job would be marked as success. Anyways you can revisit it when you implement that part.

* Function returns false instead of `nil` when there is any error. (similary to active record save)
* validate_result should be before the exception handling

```
   result = HttpParty.get()
   return false unless valid_result?(result)
 
   result
   rescue =>e
     log
    false
```

Reply:
@junan  / @ankitkalia1195 

At first I was afraid to put the `valid_result?` method inside the begin/rescue blocks: In the past I've been told we should not try/catch the whole code, only what is likely to throw errors (like the HTTP request -- as it depends on an external system). 

But if you tell me it's OK with Rails, then awesome! 

I love the eloquence of `return false unless valid_result?(result)` ! :) 

though I have a last question: What is the difference between returning `nil` and returning `false` ? 
As if I remember, doing `if nil` is the same as `if false`, no? :) 

Thanks for your info!

Reply:
I keep this conversation opened: Code changed a lot, I've added more specs to test. You might have related comments to add in here :)

Reply:
> What is the difference between returns nil and returning false?


We return a boolean value(`true`, `false`) when the method ends with `?` like `valid?` which expects a boolean value. for other method (without `?` end) I personally return `true/false` to ensure the method's success or  failure.

> As if I remember, doing if nil is the same as if false, no? :)

In this case they are same though they are different data type. Like `nil == false` will return `false`

Reply:
Thanks, very clear explanation!! :)